File: week01_08_10/BOJ_9663_정준호.py

# 8월 1주차
# https://www.acmicpc.net/problem/9663

global result

def perm(index):    # index : 다음 행
    global result, queen
    if(index == N):
        result += 1
        return

    for next in range(N):   # cnt 행에 퀸 들어갈 위치
        flag = True
        for i in range(index):
            if next == queen[i]:   # 같은 열이 있으면
                flag = False
                break
            elif next == queen[i] + (index-i):    # 대각선 체크1
                flag = False
                break
            elif next == queen[i] - (index-i):    # 대각선 체크2
                flag = False
                break
        if(flag):
            queen.append(next)
            perm(index+1)
            queen.pop()


N = int(input())
result = 0
queen = []
for i in range(N):
    queen.clear()
    queen.append(i)
    perm(1)     # 1 행의 퀸 위치 0~N-1
print(result)

File: week01_08_10/BOJ_9663_정준호.java

import java.util.Arrays;
import java.util.Scanner;

public class Main_G4_9663_NQueen {
	
	static int N;
	static int[] queen;
	static boolean[] check;
	static int count=0;
	
	public static void main(String[] args) {

		Scanner sc = new Scanner(System.in);
		N = sc.nextInt();
		queen = new int[N];
		check = new boolean[N];
		perm(0);
		System.out.println(count);
	}
	
	public static void perm(int cnt) {
		if(cnt==N) {
			count++;
			return;
		}
		
		for(int i=0; i<N; i++) {
			boolean flag = false;
			if(check[i]==true)	continue;	// 같은 열 체크
			for(int q=cnt-1, k=1; q>=0; q--, k++) {		// 직전 행 퀸 배치까지 검열
				if(cnt==0)	break;
				if(queen[q] == i-k) {	// 대각선 체크1
					flag = true;
					break;
				}
				if(queen[q] == i+k) {	// 대각선 체크2
					flag = true;
					break;
				}
			}	
			if(flag) continue;
			
			// cnt행 i열 -> 퀸 배치
			queen[cnt] = i;
			check[i] = true;
			perm(cnt+1);
			check[i] = false;
			
		}
	}
}
